packages/framework/src/resources/workflow/workflow.resource.test-d.ts

```diff
@@ -0,0 +1,281 @@
+      workflow('without-schema', async ({ step }) => {
+        await step.email('without-schema', async (controls) => {
+          expectTypeOf(controls).toEqualTypeOf<Record<string, unknown>>();
+
+          return {
+            subject: 'Test subject',
+            body: 'Test body',
+          };
+        });
+      });
+    });
+
+    it('should infer an unknown record type in the skip function step controls', async () => {
+      workflow('without-schema', async ({ step }) => {
+        await step.email(
+          'without-schema',
+          async (controls) => {
+            return {
+              subject: 'Test subject',
+              body: 'Test body',
+            };
+          },
+          {
+            skip: (controls) => {
+              expectTypeOf(controls).toEqualTypeOf<Record<string, unknown>>();
+
+              return true;
+            },
+          }
+        );
+      });
+    });
+
+    it('should infer an unknown record type in the provider function step controls', async () => {
+      workflow('without-schema', async ({ step }) => {
+        await step.email(
+          'without-schema',
+          async (controls) => {
+            return {
+              subject: 'Test subject',
+              body: 'Test body',
+            };
+          },
+          {
+            providers: {
+              sendgrid: async ({ controls }) => {
+                expectTypeOf(controls).toEqualTypeOf<Record<string, unknown>>();
+
+                return {
+                  ipPoolName: 'test',
+                };
+              },
+            },
+          }
+        );
+      });
+    });
+
+    it('should infer an unknown record type in the workflow event payload', async () => {
+      workflow('without-schema', async ({ step, payload }) => {
+        await step.email('without-schema', async () => {
+          expectTypeOf(payload).toEqualTypeOf<Record<string, unknown>>();
+
+          return {
+            subject: 'Test subject',
+            body: 'Test body',
+          };
+        });
+      });
+    });
+
+    it('should infer an unknown record type in the workflow event controls', async () => {
+      workflow('without-schema', async ({ step, controls }) => {
+        await step.email('without-schema', async () => {
+          expectTypeOf(controls).toEqualTypeOf<Record<string, unknown>>();
+
+          return {
+            subject: 'Test subject',
+            body: 'Test body',
+          };
+        });
+      });
+    });
+
+    it('should infer an unknown record type in the custom step results', async () => {
+      workflow('without-schema', async ({ step }) => {
+        const result = await step.custom('without-schema', async () => {
+          return {
+            foo: 'bar',
+          };
+        });
+
+        expectTypeOf(result).toMatchTypeOf<Record<string, unknown>>();
+      });
+    });
+  });
+
+  describe('json-schema', () => {
```

Comment:
Json schema is used as the baseline schema for asserting on inferred schema usage within the `workflow` API. This provides an adequate level of integration testing as other and future schemas are tested only with the `validateData`, `transformSchema`, `FromSchema`, and `FromSchemaUnvalidated` APIs, which as an integration point for schema validation and inference, respectively.